### ISSUE TITLE
ligatures and syllabation

### ISSUE THREAD

**Issue by eroux** [gregorio-project/gregorio#1098]
The word *efficiamur* is usually syllabized (is that English?) as `ef-fi-ci-a-mur`, but with fonts containing an `ffi` ligature, the "gabcization" depends on the context:

* if the notes are small enough not to add hyphens between the first two syllables, it should be `e()ffi()ci()a()mur()` in order to get the `ffi` ligature correctly
* if an hyphen must be put between the first two syllables, it should be `ef()fi()ci()a()mur()`

I have no clue on the best way to handle this in a simple way... do you?

**Comment by eroux** [comment]
Well, thinking about it it may not be that complex... I think the following algorithm should work:

* have a function listing all the ligatures of the font (I think that's the most difficult part), or just let user do this
* if an hyphen is needed between two syllabes, don't do anything
* if not, then detect if the last letter forms a ligature with the first letter of the next syllable:
   * if so, don't output it and keep in mind that it should be added to the next syllable
   * if not, do nothing

This is relatively simple and works fine. The big problem is with end of lines (because we shouldn't have `e` on one line and `ffi` on the next one). For this I admit I don't really know how to do... It certainly could be solved with discrectionaries, but it's super complex...

**Comment by henryso** [comment]
Unrelated to this issue per se, but I was curious so I looked it up.  The English word is syllabicate (-> syllabication) or syllabify (-> syllabification).  Syllabicate is mid-1600s, Syllabify is mid-1800s.

**Comment by henryso** [comment]
The font table internals are still undocumented and subject to change, correct?  An initial probe shows that different types of fonts store ligatures differently in the font tables, so I would be very wary of spending a lot of time reverse engineering the ligature representation in the font tables for them to change.

**Comment by eroux** [comment]
Sure, don't spend too muche time on that, this could be user-defined in a first time (only advanced users need that feature anyway). Then I think the best way to do that would be to ask luaotfload to provide a high-level function for that. Also, thanks for the vocabulary, actually "syllabation" doesn't sound French ("syllabisation" would sound more French), so I always tend to think it's English!

**Comment by eroux** [comment]
Something interesting: if we make a list of common ligatures that could be separated by an hyphen (I think the list is just `ff`, `fl` and `tt`, as `ffi` and `ffl` will be handled by the `ff` case, and there are no case in which `fi` would be cut), it's enough. If some of these ligatures are not present in the font, this is not a problem, as in both cases, the result will be fine, it's just that the burden was useless. So inspecting the font internals is really not worth it... What do you think?

**Comment by eroux** [comment]
Another issue I didn't think about: In the usual case where the notes are centered above the vowel, there shouldn't be any problem (as no ligature starts with a vowel), but in the case where the notes are centered on the whole syllable, the rendering may be a bit weird in some cases. I'm not sure there's an easy way to deal with that, but we can keep that for later...

**Comment by henryso** [comment]
I think the ones you listed are OK for the first cut, but font designers have the ability to go really nuts with ligatures, so eventually, we want to get the list from the font.

**Comment by eroux** [comment]
I took a look at the fonts some big projects are using, the only ligatures that don't fit it the list are ligatures that won't appear in a Latine text: `fh`, `fj`, `fk`, `ft` and  `fb`, so I still think `ff`, `tt` and `fl` will cover 99% of cases (looking again, maybe `ft` would be good too...). But of course, once LuaTeX 1.0 will be released and the font loader will start to be stable and documented, we certainly should get it from the font itself.

**Comment by henryso** [comment]
Does font kerning (meaning the kerning tables in the fonts themselves) work properly across syllable boundaries?

**Comment by eroux** [comment]
It would be nice, but I don't think they do... LuaTeX inherited a mechanism from Omega to do that, called `\leftghost` and `\rightghost`, but I'm not sure they work fine, and I'm quite sure they don't work fine with luaotfload... It's been on my todolist for at least 6 years, I never managed to make it a priority, but would probably be the proper time!

**Comment by henryso** [comment]
As I look at this, I'm wondering if there's a better way to do this, perhaps manipulating nodes from a lua callback?

**Comment by eroux** [comment]
Do you mean for kerns? Yes, I think `\leftghost` and `\rightghost` should be reimplemented in a generic way (in another package I mean, certainly with new names) using lua node manipulation.

**Comment by henryso** [comment]
Actually, I was considering options for ligatures. If we can tell that some multi-syllable portion of a word is continuous, we could theoretically replace the entire portion with a box having all the letters, thus allowing LuaTeX to apply ligatures and kerning.

Of course, there are tricky edge cases, like when TeX applies a stretch between syllables.  And that's only if this kind of manipulation is even possible.

**Comment by eroux** [comment]
Even if that would be possible, this can be tricky: if you have an `ii` ligature in the font, and you have `i(g)i(g)`, if you have no hyphen between the two (which should almost never be the case... but anyway...), how will you center the notes? The problem could be avoided by never ligaturing the vowels, I think this could work... But I think applying this could be almost impossible when the centering is done on the whole syllable... Unless you have ideas? This shouldn't be blocking though, we can just not use this behavior in this case.

**Comment by henryso** [comment]
I can think conceptually how I'd want to approach it:

- align the way we do now
- make all the text in the boxes containing continuous letters invisible (but still maintain their width)
- overlay a box containing all the characters

I realize there will be issues at the microtype level (ligatures having different width that the characters they replace, kerning, TeX rubber spacing, etc.).

However, I really don't know if any of this is even possible before thinking about the edge cases in more detail.

**Comment by eroux** [comment]
Yes, what I'm talking about is mainly ligatures with different width... But really, I don't there's any way we can do this in a very generic way for "english" centering without doing super convoluted stuffi like measuring ligatures, giving ratio so letters, etc. which would require some very fragile code... I think if we find a solution that works for the general case for vowel centering, that's really awesome!

Even with vowel centering, there would be many hypothetical cases in which it wouldn't work, for instance if you have a score with `eff()i()` and an `ffi` ligature in the font, things start to get ugly... but the good thing is that this should never happen in Latine, which is the primary target language.

**Comment by henryso** [comment]
I wasn't even considering syllable vs. vowel centering.  I was just trying to find a less kludgy way of doing it than transplanting a letter, which in and of itself could be tricky when styling is involved.

**Comment by eroux** [comment]
Oh, sure, if there's any styling involved, I think this feature can be forgotten, I don't think it's doable in a completely generic way, I think it could a small set could be done first, then things could be added case by case...

**Comment by henryso** [comment]
Since my previous attempt is apparently not viable, why not just put all post-center letters into the next syllable when a hyphen is not necessary?  It's still not perfect and will fail in some cases when the centered letters are part of a ligature, but it would eliminate (for now?) the need to match ligatures.  Any downsides to this approach (yes, I realize there will be problem at the end of lines, as described above).

**Comment by eroux** [comment]
That's a bit heavy, but why not yes, apart from the end of line problem (which is not specific to this approach), I think this would work just fine!

**Comment by henryso** [comment]
The end-of-line case may be the really difficult part here.  The end-of-line hyphen (if it's necessary) is added in the Lua code by the post line break filter callback.  I'm assuming that putting the text in an `\hbox` will prevent ligaturing through the box, so the question becomes how to move the letters to the next syllable if the end-of-line hyphen is *not* added by the Lua code.

**Comment by eroux** [comment]
I'm not entirely sure I understand what you have in mind, can you give a more precise example?

**Comment by henryso** [comment]
The idea is to take two syllables like:

`((before1)(center1)(after1))((before2)(center2)(after2))`

and transform it to

`((before1)(center))((after1)(before2)(center2)(after2))`

when no hyphen is necessary.

However, if `((before1)(center1)(after1))` is at the end of a line, the Lua code would put a hyphen after it, and we would not want to alter it.

So, we need to figure out how to move `(after1)` into the second syllable when the first syllable is *not* at the end of a line (thus the Lua code does not add a hyphen).

There are two options I can think of:

- detect the situation (somehow) and do the transformation in TeX by twiddling the arguments.
- do the transformation (somehow) in Lua while processing the post line break filter and the syllable is not at the end of the line.

In both of these cases, you know (by construction) how do to one thing, but not how to do the other.

**Comment by eroux** [comment]
I see. Well, I think we already more or less have a mechanism to detect end of lines and feed it to TeX, maybe we could reuse that? (I don't have all of this in mind, and I won't be very available this week-end, so I'm affraid I'm not very helpful, sorry!)

**Comment by henryso** [comment]
We sort of have that, but it involves recording positions on the PDF generation pass, and using that mechanism here will likely result in an unstable document.

**Comment by eroux** [comment]
Indeed, that might be unstable in some edge cases (though they should be quite rare). Well, let's wait for your idea then, thanks!

**Comment by henryso** [comment]
My idea is to number the syllables using whatsit markers and record the syllable numbers at the end of lines.  However, I think this still can produce an unstable document because ligatures and kerning can cause TeX to break the line differently.

That aside, I have one more technical issue.  My idea would have GregorioTeX make no changes on the first pass so that line endings can be computed correctly and recorded.  In the second pass, GregorioTeX would use the information from the first pass to move the final part of the syllable to the next syllable.  If the user makes a change to the .gabc or .tex document at this point, the first run will start moving the text around because the previous run has been recorded in the .gaux file.  When applying the end-of-line hyphens, the Lua code will realize something has changed, invalidating this run; however, by that point it's too late to compute line endings.  Therefore, *two more* runs are required to finally get the positioning right.  Is there a way to detect a change on the first run before anything has happened so only two runs are then required?

I'm also soliciting ideas about the document stability thing.  Maybe I should force a line break on later passes (this assumes that ligatures and kerning make more room for text, which is typical)?  However, this interacts with the ability of GregorioTeX to detect a "real" change in line breaking, which means that it could produce a different result (after some runs) with and without the .gaux file.

**Comment by eroux** [comment]
At first I thought moving the syllable in lua was almost impossible becase messing up with the text in lua would move the text/notes alignment. But remembering how things are done, I'm quite sure it wouldn't move it, so it should be doable to move the syllables in lua. This would certainly be costy though...

Here are two more approaches about the stability problem:

1. we could compute a kind of hash at the beginning of a score, containing the font id, the font size, and the different spaces, so that a change in these could be detected easily. Maybe the hash of the spaces could be computed at each spacing change in `grecreatedim` and `grechangedim`, so that it doesn't have to be computed at every score

2. we could also record the positions (and hash) that were recorded two runs ago, and if we see that we're just going back to the previous positions with no hash change, then we stop changing. This is a bit fragile though...

I don't really like the idea of locking the line breaks, but why not...

**Comment by henryso** [comment]
I'd like to understand the movement done in Lua a little better, as that might end up being cleaner and prevent the problems during multiple runs.  Where should I look?

**Comment by eroux** [comment]
That's just node manipulation plus applying the ligaturing and kerning functions... I'm affraid I cannot say much more... it won't be very easy and straightforward I'm affraid...

**Comment by henryso** [comment]
OK, since ligaturing and kerning occur before line breaking, attempting to move the text in Lua is no less "clean" than moving it in TeX, so I'll go back to my original idea.

On reducing the number of runs, I like the idea of hashing.  I will think about how to implement that.  I think the extra step of recording the positions is a bit too fragile, so my first attempt will not include it.

I agree about locking the line breaks.  I'm not satisfied with that idea, so I'm still thinking about it.

**Comment by henryso** [comment]
I pulled some changes into a fix-1098 branch, which is the start of my attempt to get this working.  I'm having some trouble figuring out the fix to a problem I have in the code.

Note: the changes so far don't include the hashing idea described by @eroux.  It does move everything after the center to the next syllable unless the syllable is at the end of a line.  However, this seems to cause TeX to put some hyphens in where it didn't before.  See for example the FactusEst test:

release-4.2 branch:
![image](https://cloud.githubusercontent.com/assets/421318/15277442/4bd65860-1ad2-11e6-9169-5b4c73fbeea5.png)

fix-1098 branch
![image](https://cloud.githubusercontent.com/assets/421318/15277440/29d74292-1ad2-11e6-8962-81739b4bb248.png)

Note the change in "Factus", "advenientis", "vehementis".  However, the algorithm is *not* completely broken and still works for "repente", "repleti", "omnes", "loquentes", and "magnalia".

I'd appreciate it if someone can take a look and see if they have any ideas.

**Comment by eroux** [comment]
Thanks! I cannot see anything obvious casing the issue, I'm not really able to debug today, I'll try to take a look tomorrow

**Comment by henryso** [comment]
I edited my earlier comment to strike out some words for which I thought the algorithm was working.  They actually *don't* indicate any sort of success because of the way the syllables are entered in the test, and "repente" is actually an indication of failure.  So there are far more failures than successes.

In additional testing, I have found that the text does properly get merged if the post-center part is manually placed in the gabc (i.e., `FA(..)ctus(...)` vs `FAc(...)tus(...)`), so there must be something wrong with how I am moving the syllable in TeX.  I haven't been able to figure out what that is.

**Comment by henryso** [comment]
The problem is that the `nextbegindifference` calculation was not taking into account the post-center part that would be moved to the next syllable.  I can work further based on this finding, but I may need some help as the cases become more eccentric.

**Comment by henryso** [comment]
I don't think the calculations take fixed text styles into account.  If someone can confirm this, we should create a new issue to track it.

**Comment by rpspringuel** [comment]
The calculation of `nextbegindifference` uses `\gre@calculate@textaligncenter` to compute the portion of the distance based on the text and that macro does account for the text styles.  If you're seeing some indication that it's not doing so correctly, then there might be a bug in `\gre@calculate@textaligncenter`.

**Comment by henryso** [comment]
I'm probably wrong, then, but this might be why it's acting so weird with my changes.

**Comment by henryso** [comment]
I made some changes (pushed to the `fix-1098` branch) which fix the problem I mentioned earlier with too much hyphenation.  However, now there is a problem where the space after the notes above `den` in `sedentes` and above `quen` in `loquentes` is lost.

![image](https://cloud.githubusercontent.com/assets/421318/15451571/cc67db7e-1f97-11e6-9873-e2ab6faaf0ae.png)

These are points where GregorioTeX *is* adding a hyphen, so the code does not shift any text around.  I haven't been able to figure out what is supposed to be adding a space there.  Does anyone have any hints for places to look?

**Comment by rpspringuel** [comment]
I think that space that goes in the places indicated is `\gre@skip@syllablefinalskip`, which, naturally enough, is calculated by `\gre@calculate@syllablefinalskip` (defined in `gregoriotex-spaces.tex`), though the actual insertion is in `\GreSyllable` (in `gregoriotex-syllable.tex`).

**Comment by henryso** [comment]
Thanks, I was able to track down the issue and took your hint from the earlier commit to consolidate the formatting.  At this point, I think I will need to defer to you before I can progress further.

There are two differences that I see running the tests.  One of them is a small shift of the neume in the illumination and hyphenation tests.  Obviously some computation is off, but I don't know which one:

![image](https://cloud.githubusercontent.com/assets/421318/15451936/8e2e4ab6-1fa9-11e6-80e6-e2612ba0545c.png)

The second difference is some odd behavior in the bar spacing tests: bar-vs-nobar, bar-spacing-new, and bar-spacing-old.  I don't understand the bar spacing algorithms well enough and the test results are a little hard to follow.

Finally, fix-396 is failing to typeset, most likely because of instability in the output; nothing has been coded to handle this known issue, so I'm ignoring that failure for now.

**Comment by rpspringuel** [comment]
Clivis alignment perhaps?  If the first two notes are interpreted as a clivis (and the clivis alignment rules are in play), then their alignment point would be in the center of them, which is what it looks like the blue notes are doing.  If they are not a clivis (or the rules are turned off), then the center of the first note would be the alignment point, which is what the red notes appear to be doing.  My guess then is that something is changing the classification of the neume from not being considered a clivis to being considered one.

Since this is in the illumination and hyphenation tests the gtex file doesn't get checked automatically, but I'd suggest that a manual comparison of the before and after gtex file for those tests is in order.

**Comment by henryso** [comment]
I think we can safely rule out gtex file generation. No C code has changed in this branch versus release-4.2.

**Comment by rpspringuel** [comment]
Okay.  Well, looking at the gabc for that word, it is coded as `vehe(h_f!gwh)men(hi~)tis,(g)`.  This means we're dealing with a syllable which splits into `v`, `e`, and `he` and which does start with a clivis (`h_f`).  Checking the code, I think I see the issue.  The default clivis alignment setting decides how to align the clivis based on the length of the after vowel portion of the syllable.  Your code must be changing the perceived length of the that portion of the syllable and thus switching the alignment decision.  Of course, I don't know why that might be the case, as there should be no ligatures between `he` and `m`.

**Comment by henryso** [comment]
That actually makes complete sense, as the code would move `he` to the next syllable.  Now I have to figure out how to fix it.  Thanks!

**Comment by henryso** [comment]
With that resolved, the remaining formatting issue is differences that occur when a bar syllable with text is not a separate word.  For example `{3}3333333333(g)*(;){}4444(gxgf)` (from the end of `bar-spacing.gabc`) essentially renders as `{3}(g)3333333333*(;){}4444(gxgf)` which throws the formatting off.  My inclination is to prevent text relocation into a bar syllable.  Any objections?

**Comment by eroux** [comment]
Sorry, I'm just coming back from a few days offline, no objection to your proposal!

**Comment by henryso** [comment]
I've pushed my latest changes into the branch.  I also accidentally included an early (and quite broken) attempt to stabilize the output of the fix-396 test, but decided it may be beneficial to leave the code there (for now) for the purposes of discussion.  I currently have no good solution to the stability problem.
